sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableOperations.scala

```diff
@@ -144,19 +144,33 @@ case class ParquetTableScan(
         new Iterator[Row] {
           def hasNext = iter.hasNext
           def next() = {
-            val row = iter.next()._2.asInstanceOf[SpecificMutableRow]
+            iter.next()._2 match {
+              case row: SpecificMutableRow => {
 
-            // Parquet will leave partitioning columns empty, so we fill them in here.
+              // Parquet will leave partitioning columns empty, so we fill them in here.
             var i = 0
             while (i < requestedPartitionOrdinals.size) {
-              row(requestedPartitionOrdinals(i)._2) =
-                partitionRowValues(requestedPartitionOrdinals(i)._1)
+              row (requestedPartitionOrdinals (i)._2) =
+              partitionRowValues (requestedPartitionOrdinals (i)._1)
               i += 1
             }
             row
+            }
+              case row : Row => {
+                val rVals = row.to[Array]
+                var i = 0
+                while (i < requestedPartitionOrdinals.size) {
+                  rVals
+                    .update(
+                      requestedPartitionOrdinals (i)._2,
+                      partitionRowValues (requestedPartitionOrdinals (i)._1))
+                  i += 1
+                }
+                Row.fromSeq(rVals)
+              }
           }
         }
-      }
+      }}
```

Comment:
For indentation and whitespace related styling, please refer to equivalent changes in #4792.